torch/_inductor/sizevars.py

```diff
@@ -384,52 +389,86 @@ def wrapper(*args, **kwargs):
+
+        stride_symbols = [
+            sympy.Wild(f"stride{i}", exclude=vars) for i in range(len(var_symbols))
+        ]
+        var_to_stride = {v: s for v, s in zip(var_symbols, stride_symbols)}
+        offset_symbol = sympy.Wild("offset")
+        index_pattern = offset_symbol + sympy_dot(var_symbols, stride_symbols)
+
+        match = index.match(index_pattern)
+        if match is None:
+            # Index calculation is not strided
+            return [None] * len(vars), None
+
+        strides = []
+        for v in vars:
+            if v not in var_to_stride:
+                stride = 0 if v == 0 else None
             else:
-                # TODO(jansel): should we use sympy.diff here?
-                strides.append(
-                    sympy_subs(index_dim, {v: sympy.Integer(1)})
-                    - sympy_subs(index_dim, {v: sympy.Integer(0)})
-                )
+                stride = match.get(var_to_stride[v], None)
+            strides.append(stride)
+
+        offset = match[offset_symbol]
+
+        # If any vars appear in the offset terms, they are not strided
+        vars_set = set(vars)
+        if vars_set & offset.free_symbols:
+            for i, v in enumerate(vars):
+                if v in offset.free_symbols:
+                    strides[i] = None
+
+            offset = None
+
+        return strides, offset
+
+    def maybe_stride_vars(
+        self, index: Expr, vars: List[sympy.Symbol]
+    ) -> List[Optional[Expr]]:
+        """Convert an indexing expression back into strides"""
+        strides, offset = self.maybe_stride_and_offset_vars(index, vars)
         return strides
 
-    def offset_var(self, index: Expr, vars: List[sympy.Symbol]) -> Expr:
+    def maybe_offset_var(self, index: Expr, vars: List[sympy.Symbol]) -> Optional[Expr]:
         """Extract offset part of an indexing expression"""
-        index = self.simplify(index)
-        return sympy_subs(index, {v: sympy.Integer(0) for v in vars if v != 0})
+        strides, offset = self.maybe_stride_and_offset_vars(index, vars)
+        return offset
 
     def stride_hints(self, index: Expr, vars: List[sympy.Symbol]) -> List[int]:
         for v in index.free_symbols:
             if v.name.startswith("indirect"):
                 index = sympy_subs(index, {v: 0})
         result = []
-        for s in self.stride_vars(index, vars):
+        for i, s in enumerate(self.maybe_stride_vars(index, vars)):
+            if s is None:
+                # If the dimension is not truly strided, we approximate a
+                # stride by taking the difference of the 1st and 0th element in
+                # that dimension.
+                zero_index = {v: sympy.Integer(0) for v in vars if v != 0}
+                first_index = dict(zero_index)
+                first_index[vars[i]] = sympy.Integer(1)
+                s = sympy_subs(index, dim_only) - sympy_subs(index, zero_index)
```

Comment:
This fixes the regression for me locally. `decide_layout()` was using the stride_hint of zero to always put the `ModularIndex` dimensions first. In that context I think using the old method is okay since you shouldn't be relying on `stride_hint` for correctness anyway.